### PR TITLE
Add python 3.8 compatibility

### LINKED ISSUE #277: Ensure Python 3.8 compliance, update docs and tests to reflect that

### PR DESCRIPTION
**Pull request by dfrancisco1998**
#### What does this implement/fix? Explain your changes.
Closes #277

### CODE REVIEW

**Comment by dfrancisco1998** [comment]
@bdpedigo I have tried to change the docstring tests in order to accommodate what I saw the desired output array change due to the update(a change which I am having a lot fo trouble tracking down that is causing such minute differences). If you look at the travis output for this you will see in python 3.8 I pass the rpdg_corr.py but I do not pass those tests in python 3.6 and 3.7. The output that 3.6 and 3.7 get is innately different than the output array that 3.8 is supposed to get and I do not know how to conciliate between the different versions because each version desires to have its own specific output array. I had been scratching my head at this for the weekend and had made a few PR requests(that I have closed because they were nothing worthwhile). Any guidance on what to do next would be appreciated. thanks!

**Comment by bdpedigo** [comment]
> @bdpedigo I have tried to change the docstring tests in order to accommodate what I saw the desired output array change due to the update(a change which I am having a lot fo trouble tracking down that is causing such minute differences). If you look at the travis output for this you will see in python 3.8 I pass the rpdg_corr.py but I do not pass those tests in python 3.6 and 3.7. The output that 3.6 and 3.7 get is innately different than the output array that 3.8 is supposed to get and I do not know how to conciliate between the different versions because each version desires to have its own specific output array. I had been scratching my head at this for the weekend and had made a few PR requests(that I have closed because they were nothing worthwhile). Any guidance on what to do next would be appreciated. thanks!

What have you been trying to do w.r.t. tracking down the difference between the two? 

A few half-formed ideas: 
- this is simulation code, which means that the result is random, but over many random samples, the simulation must have certain properties. have you checked whether this is just some difference caused by random sampling or whether the sampled graphs (in python 3.8) are what we'd expect (i.e. do they still follow the expected distribution)?
- have you looked for (googled) differences between random number generation in numpy between python 3.7 and 3.8?
- have you tried setting the same random seed in 3.7 and 3.8 and sampling random numbers to make sure they are the same? 
- (related to the one above) but if you don't understand why the results are different overall, I would break down into smaller problems, and check whether intermediate values in the underlying code to do the sampling are the same, and look for where the results diverge

**Comment by loftusa** [comment]
relevant to this in case you didn't see it already: https://docs.python.org/3/whatsnew/3.8.html#porting-to-python-3-8

**Comment by dfrancisco1998** [comment]
> > @bdpedigo I have tried to change the docstring tests in order to accommodate what I saw the desired output array change due to the update(a change which I am having a lot fo trouble tracking down that is causing such minute differences). If you look at the travis output for this you will see in python 3.8 I pass the rpdg_corr.py but I do not pass those tests in python 3.6 and 3.7. The output that 3.6 and 3.7 get is innately different than the output array that 3.8 is supposed to get and I do not know how to conciliate between the different versions because each version desires to have its own specific output array. I had been scratching my head at this for the weekend and had made a few PR requests(that I have closed because they were nothing worthwhile). Any guidance on what to do next would be appreciated. thanks!
> 
> What have you been trying to do w.r.t. tracking down the difference between the two?
> 
> A few half-formed ideas:
> 
> * this is simulation code, which means that the result is random, but over many random samples, the simulation must have certain properties. have you checked whether this is just some difference caused by random sampling or whether the sampled graphs (in python 3.8) are what we'd expect (i.e. do they still follow the expected distribution)?
> * have you looked for (googled) differences between random number generation in numpy between python 3.7 and 3.8? I doubt there is anything here, but worth checking whether other people have found similar issues
> * have you tried setting the same random seed in 3.7 and 3.8 and sampling random numbers to make sure they are the same?
> * (related to the one above) but if you don't understand why the results are different overall, I would break down into smaller problems, and check whether intermediate values in the underlying code to do the sampling are the same, and look for where the results diverge

@bdpedigo in order to track down where there could have been a divergence I have made jupyter notebook that pretty much calls all the random aspects and checks them. I am having a little trouble understanding how to check if a graph is following the distribution properly ( I am still trying to learn all these new concepts of graphs etc). From what I have gathered (and can be seen here: https://github.com/dfrancisco1998/ndd_stuff/blob/master/3.8%20compliance%20travis%20vs%20notebook%20check%20.ipynb) running the code that causes the problem in a notebook is fine. This is why I did not make a python 3.7 env because everything seems to be acting fine in my notebook. 

The env I am using is: networkx   2.5, numpy 1.19.1 , matplotlib 3.1.1, pandas 1.1.1, pytest 6.0.1, pytest-cov 2.10.1, python  3.8.5, scikit-learn  0.23.2,scipy 1.5.0, seaborn  0.10.1. 

I am just confused about how this is possible and am very confused about where the actual problem is coming from. 

Again I apologize for asking so many questions, I am just confused. Thanks for your patience!

**Comment by dfrancisco1998** [comment]
@bdpedigo I have closed all other PRs and only have this one open. I have also turned the docstring tests into a python script(the same code that was in jupyter notebook attacked above) . 

https://github.com/dfrancisco1998/ndd_stuff/blob/master/test_travis_simulations_rdpg.py

https://github.com/dfrancisco1998/ndd_stuff/blob/master/test_travis_rdpg.py

Hope this helps!

File: graspy/simulations/simulations.py [hunk 1]
```diff
@@ -695,7 +695,7 @@ def rdpg(X, Y=None, rescale=False, directed=False, loops=False, wt=1, wtargs=Non
            [0., 0., 0., 1., 1.],
            [0., 1., 1., 0., 0.],
            [1., 1., 1., 0., 0.]])
-
+
```
Comment (bdpedigo) [inline graspy/simulations/simulations.py:698]: @dfrancisco1998 what is the point of this change? just making sure it doesnt mess up doctest or anything

Comment (dfrancisco1998) [reply graspy/simulations/simulations.py:698]: This is the correct doctest (same one from master). Sorry I was trying to correct the edits I had made in the file in my branch and must have added in an extra space.

Comment (bdpedigo) [reply graspy/simulations/simulations.py:698]: mind removing this change then?

**Comment by bdpedigo** [comment]
@dfrancisco1998 please also run black on all files, right now CI is failing on python 3.6/3.7 but I think this is just cause of black?

**Comment by bdpedigo** [comment]
Screen Shot 2020-09-18 at 9 27 29 AM

@dfrancisco1998 is this what you were expecting to see?

**Comment by bdpedigo** [comment]
@dfrancisco1998 can you also remind me the current state of affairs (sorry for being dense but I am juggling many different PRs so I'd like to have this written down in one place). Your current assertion is: 
- all tests/doctests pass when you run pytest/doctest locally on your machine with python3.8
- some doctests fail when you add 3.8 to travis and travis runs pytest/doctest.
is the above correct? if not, please correct me

**Comment by dfrancisco1998** [comment]
> @dfrancisco1998 can you also remind me the current state of affairs (sorry for being dense but I am juggling many different PRs so I'd like to have this written down in one place). Your current assertion is:
> 
> * all tests/doctests pass when you run pytest/doctest locally on your machine with python3.8
> * some doctests fail when you add 3.8 to travis and travis runs pytest/doctest.
>   is the above correct? if not, please correct me
You are totally good. There must be a lot going on with the PRs and shift with Microsoft. And yes oddly I can run the code on my local machine and it outputs what is desired by the doctest (please check jupyter notebooks and py files linked above) as denoted in the docstring form master(as well as my current PR). But when travis runs this for some reason those two tests fail and not output the same thing as desired by the doctest. Hope this clarifies!

**Comment by dfrancisco1998** [comment]
> Screen Shot 2020-09-18 at 9 27 29 AM
> 
> @dfrancisco1998 is this what you were expecting to see?

Yes this is what should be outputted but has not been with travis

**Comment by dfrancisco1998** [comment]
> @dfrancisco1998 please also run black on all files, right now CI is failing on python 3.6/3.7 but I think this is just cause of black?

Would you like me to run black and PR again?

**Comment by bdpedigo** [comment]
> Would you like me to run black and PR again?

Yes but don't need to make a new PR. When you push to this branch it'll just update this PR

**Comment by bdpedigo** [comment]
@dwaynepryce is it true that in #471 you are running doctest as part of pytest ([this line](https://github.com/microsoft/graspologic/blob/0d70ac1e8f46336fb3250ecab2d313be064f1b3d/pytest.ini#L2)) and then pytest (and doctest) get run on the matrix of python [3.6 3.7 3.8] and [mac windows linux] ([this line](https://github.com/microsoft/graspologic/blob/0d70ac1e8f46336fb3250ecab2d313be064f1b3d/.github/workflows/build.yml#L58))? And everything pytest/doctest appears to be passing?

Reason I ask is we are having some weird issues on this PR where everything in doctest looks fine locally but not when it gets run on travis. but we're kicking travis to the curb anyway so maybe we shouldn't fret about it.

**Comment by bdpedigo** [comment]
> @dwaynepryce is it true that in #471 you are running doctest as part of pytest ([this line](https://github.com/microsoft/graspologic/blob/0d70ac1e8f46336fb3250ecab2d313be064f1b3d/pytest.ini#L2)) and then pytest (and doctest) get run on the matrix of python [3.6 3.7 3.8] and [mac windows linux] ([this line](https://github.com/microsoft/graspologic/blob/0d70ac1e8f46336fb3250ecab2d313be064f1b3d/.github/workflows/build.yml#L58))? And everything pytest/doctest appears to be passing?
> 
> Reason I ask is we are having some weird issues on this PR where everything in doctest looks fine locally but not when it gets run on travis. but we're kicking travis to the curb anyway so maybe we shouldn't fret about it.

Ah I see, after @j1c's fix it looks like the same issue is in #471

**Comment by bdpedigo** [comment]
closing, we'll figure out 3.8/docstring later

**Pull request closed without merging**